### PR TITLE
perf(useErrorBoundary): assert parent ErrorBoundaryContext following typescript way to remove unnecessary optional chaining

### PR DESCRIPTION
**Pull request by manudeli**
**What**: in useErrorBoundary, I assert parent `ErrorBoundaryContext` with useful typescript way

**Why**: to remove unnecessary optional chaining, if no optional chaining, built source code will be shorter than before. and I could guess its performance will be slightly better than before. 

### AS-IS: source code built before removing optional chaining

```ts
function assertErrorBoundaryContext(value) {
  if (value == null || typeof value.didCatch !== "boolean" || typeof value.resetErrorBoundary !== "function") {
    throw new Error("ErrorBoundaryContext not found");
  }
  return true;
}

function useErrorBoundary() {
  const context = react.useContext(ErrorBoundaryContext);
  assertErrorBoundaryContext(context);
  const [state, setState] = react.useState({
    error: null,
    hasError: false
  });
  const memoized = react.useMemo(() => ({
    resetBoundary: () => {
      context === null || context === void 0 ? void 0 : context.resetErrorBoundary(); // this code line have unncessary code because of assertion function
      setState({
        error: null,
        hasError: false
      });
    },
    showBoundary: error => setState({
      error,
      hasError: true
    })
  }), [context === null || context === void 0 ? void 0 : context.resetErrorBoundary]); // this code line have unncessary code because of assertion function
  if (state.hasError) {
    throw state.error;
  }
  return memoized;
}
```

### TO-BE: source code built after removing optional chaining

```ts
function assert(condition, messageOnError) {
  if (!condition) {
    throw new Error(messageOnError);
  }
}

function useErrorBoundary() {
  const contextValue = react.useContext(ErrorBoundaryContext);
  const hasErrorBoundaryContext = contextValue !== null && typeof contextValue.didCatch === "boolean" && typeof contextValue.resetErrorBoundary === "function"; // removed unnecessary code
  assert(hasErrorBoundaryContext, "useErrorBoundary need ErrorBoundaryContext.Provider or ErrorBoundary in parent");
  const [state, setState] = react.useState({
    error: null,
    hasError: false
  });
  const memoized = react.useMemo(() => ({
    resetBoundary: () => {
      contextValue.resetErrorBoundary(); // concise code, performance will be different
      setState({
        error: null,
        hasError: false
      });
    },
    showBoundary: error => setState({
      error,
      hasError: true
    })
  }), [contextValue.resetErrorBoundary]); // concise code, performance will be different
  if (state.hasError) {
    throw state.error;
  }
  return memoized;
}
```

**How**: I used keyword `asserts` like [TypeScript docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions)

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/assertErrorBoundaryContext.ts [hunk 1]
```diff
@@ -1,15 +0,0 @@
-import { ErrorBoundaryContextType } from "./ErrorBoundaryContext";
-
-export function assertErrorBoundaryContext(
-  value: any
-): value is ErrorBoundaryContextType {
```
Comment (bvaughn) [inline src/assertErrorBoundaryContext.ts:5]: I think the main issue with the function in `main` is a syntax error on the return type. Should be
```ts
export function assertErrorBoundaryContext(
  value: any
): asserts value is ErrorBoundaryContextType {
```

**Comment by bvaughn** [comment]
Fixed in edab410

**Comment by bvaughn** [comment]
Fixed in 23a4d77

File: src/assertErrorBoundaryContext.ts (hunk 1 above)
Comment (manudeli) [reply src/assertErrorBoundaryContext.ts:5]: Actually I didn't know this syntax(asserts + is) is possible. Yours is better👍

**Pull request closed without merging**